src/ParserInterface.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace Dhii\Parser;
+
+use Dhii\Io\Resource\ReadableResourceInterface;
+use Dhii\Parser\Exception\CouldNotParseExceptionInterface;
+use Dhii\Parser\Exception\ParserExceptionInterface;
+
+/**
+ * Something that can parse a hierarchy representation into a hierarchy.
+ *
+ * From this perspective, any parsable string can be converted into an abstract
+ * syntax tree. The more obvious examples include XML and JSON. Less obvious
+ * examples include SQL, structured English, or even a `printf()` format string.
+ *
+ * It isn't always effective to have the whole string in order to start parsing.
+ * For example, the parser may want to return information about only some
+ * of the nodes represented in the source; in this case, holding the whole
+ * source string in memory can be undesirable. If only some information about
+ * the nodes should be retrieved, the problem becomes more severe, especially
+ * if only a certain maximal amount of nodes is required in the output; in this
+ * case, a very large string in memory must traditionally be parsed into a DOM
+ * first, before the desired data can be retrieved from it. This parser solves
```

Comment:
> This parser solves that problem by moving through the resource in a linear fashion ...

It doesn't actually. Even if the param doc says that `$source` can be a `ReadableResourceInterface`, this is something that the implementation needs to implement. A resource can just as easily be read from start to finish.

I would advise against documenting this paragraph and only briefly mention that the parser supports reading from resources, so that implementations can choose to be memory efficient and only read the desired amount.

Reply:
It does, actually. A resource can easily be read from start to finish. But for that, the consumer may have to make a sequence of reads, each of a specific maximal size. The implementation can wrap a string into a resource (similarly to a buffered stream) if the whole string is already known to be loaded; but the API relies on the ability to always request the resource in parts.

Reply:
What I meant is that the documentation seems to imply that the interface does something. The way it's written it's as if it has functionality.

> This parser solves that problem by moving through the resource in a linear fashion ...

My suggestion is to document the concepts, rather than documenting what you think is an ideal implementation.